content/concepts/protocols.md

```diff
@@ -0,0 +1,268 @@
+
+The first takes two arguments: a protocol id, and a handler function. If an incoming stream request sends an exact
+match for the protocol id, the handler function will be invoked with the new stream as an argument.
+
+#### Using a match function
+
+The second kind of protocol registration takes three arguments: the protocol id, a protocol match function, and the handler function.
+
+When a stream request comes in whose protocol id doesn't have any exact matches, the protocol id will be passed through
+all of the registered match functions. If any returns `true`, the associated handler function will be invoked.
+
+This gives you a lot of flexibility to do your own "fuzzy matching" and define whatever rules for protocol matching
+make sense for your application.
+
+#### Match using semver
+
+If you'd like to concurrently support a range of numbered versions, you may want to use semantic versioning (aka [semver](https://semver.org)).
+
+In go-libp2p, a helper function called [`MultistreamSemverMatcher`](https://github.com/libp2p/go-libp2p-host/blob/master/match.go) can be used
+as a protocol match function to see if an incoming request can be satisfied by the registered protocol version.
+
+js-libp2p provides a [similar match function](https://github.com/multiformats/js-multistream-select/blob/master/src/listener/match-semver.js)
+as part of [js-multistream-select](https://github.com/multiformats/js-multistream-select/)
+
+### Dialing a specific protocol
+
+When dialing a remote peer to open a new stream, the initiating peer sends the protocol id that they'd like to use. The remote peer will use
+the matching logic described above to accept or reject the protocol. If the protocol is rejected, the dialing peer can try again.
+
+When dialing, you can optionally provide a list of protocol ids instead of a single id. When you provide multiple protocol ids, they will
+each be tried in succession, and the first successful match will be used if at least one of the protocols is supported by the remote peer.
+This can be useful if you support a range of protocol versions, since you can propose the most recent version and fallback to older versions
+if the remote hasn't adopted the latest version yet.
+
+## Core libp2p protocols
+
+In addition to the protocols that you write when developing a libp2p application, libp2p itself defines several foundational protocols that are used for core features.
+
+### Common patterns
+
+The protocols described below all use [protocol buffers](https://developers.google.com/protocol-buffers/) (aka protobuf) to define message schemas.
+
+Messages are exchanged over the wire using a very simple convention called [msgio](https://github.com/jbenet/go-msgio), which prefixes binary
+message payloads with an integer that represents the length of the payload in bytes. The length is encoded as a [protobuf varint](https://developers.google.com/protocol-buffers/docs/encoding#varints)  (variable-length integer).
```

Comment:
We are not actually using msgio afaict, but we do use varint lengh-prefixed messages.